jib-cli/src/main/java/com/google/cloud/tools/jib/cli/buildfile/Layers.java

```diff
@@ -0,0 +1,179 @@
+
+import com.google.cloud.tools.jib.api.buildplan.AbsoluteUnixPath;
+import com.google.cloud.tools.jib.api.buildplan.FileEntriesLayer;
+import com.google.cloud.tools.jib.api.buildplan.FileEntry;
+import com.google.cloud.tools.jib.api.buildplan.LayerObject;
+import com.google.common.base.Predicate;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/** Class to between different layer representations. */
+class Layers {
+
+  /**
+   * Convert a layer spec to a list of layer objects.
+   *
+   * <p>Does not handle missing directories for files added via this method. We can either prefill
+   * directories here, or allow passing of the file entry information directly to the reproducible
+   * layer builder
+   *
+   * @param buildRoot the directory to resolve relative paths, usually the directory where the build
+   *     config file is located
+   * @param layersSpec a layersSpec containing configuration for all layers
+   * @return a {@link List} of {@link LayerObject} to use as part of a buildplan
+   * @throws IOException if traversing a directory fails
+   */
+  static List<LayerObject> toLayers(Path buildRoot, LayersSpec layersSpec) throws IOException {
+    List<LayerObject> layers = new ArrayList<>();
+
+    FilePropertiesStack filePropertiesStack = new FilePropertiesStack();
+    // base properties
+    layersSpec.getProperties().ifPresent(filePropertiesStack::push);
+
+    for (LayerSpec entry : layersSpec.getEntries()) {
+      // each loop is a new layer
+      if (entry instanceof FileLayerSpec) {
+        FileEntriesLayer.Builder layerBuiler = FileEntriesLayer.builder();
+
+        FileLayerSpec fileLayer = (FileLayerSpec) entry;
+        layerBuiler.setName(fileLayer.getName());
+        // layer properties
+        fileLayer.getProperties().ifPresent(filePropertiesStack::push);
+        for (CopySpec copySpec : ((FileLayerSpec) entry).getFiles()) {
+          // copy spec properties
+          copySpec.getProperties().ifPresent(filePropertiesStack::push);
+
+          // relativize all paths to the buildRoot location
+          Path rawSrc = copySpec.getSrc();
+          Path src = rawSrc.isAbsolute() ? rawSrc : buildRoot.resolve(rawSrc);
+          AbsoluteUnixPath dest = copySpec.getDest();
+
+          if (Files.isRegularFile(src)) { // regular file
+            if (copySpec.getExcludes().size() > 0 || copySpec.getIncludes().size() > 0) {
+              throw new UnsupportedOperationException(
+                  "Cannot apply includes/excludes on single file copy directives.");
+            }
+            AbsoluteUnixPath target = dest;
+            if (copySpec.isDestEndsWithSlash()) {
+              target = copySpec.getDest().resolve(src.getFileName());
```

Comment:
Just do `dest = dest.resolve(src.getFileName())` and use `dest` below?

Reply:
I would like to limit how much I change dest, rather not it be a changed var. (I should use final, but not google style)

Reply:
I think I'll just use a `?:` for this?